e2e/LongHaul/service/DeviceOperations.cs

```diff
@@ -140,14 +156,32 @@ public async Task SendC2dMessagesAsync(Logger logger, CancellationToken ct)
                         MessageId = payload.RandomId.ToString(),
                     };
 
-                    logger.Trace($"Sending message with Id {message.MessageId} to the device: {_deviceId}", TraceSeverity.Information);
+                    logger.Trace(
+                        $"Sending message with Id {message.MessageId} to the device: {_deviceId}",
+                        TraceSeverity.Information);
                     logger.Metric(TotalC2dMessagesSentCount, _totalC2dMessagesSentCount);
 
                     await _serviceClient.Messages.SendAsync(_deviceId, message, ct).ConfigureAwait(false);
                 }
+                catch (IotHubServiceException ex) when (ex.IsTransient is true)
+                {
+                    logger.Trace(
+                        $"Caught transient exception while sending a C2D message.\n{ex}\nRetrying...",
+                        TraceSeverity.Warning);
+                    // retry
+                }
+                catch (InvalidOperationException ex) when (ex.Message.Contains("Must open client"))
+                {
+                    logger.Trace(
+                        $"{ex.Message}\nOpening client to send C2D messsage...",
+                        TraceSeverity.Warning);
+                    await _serviceClient.Messages.OpenAsync(ct).ConfigureAwait(false);
+                }
```

Comment:
in case of disconnect, the `Message` subclient would be closed and throw `InvalidOperationException` after reconnect. Add handling here.

Reply:
Shouldn't the SDK's retry policy handle the reopening after disconnect? Why would the user need to make this call?

Reply:
Yeah, this behavior seems wrong.

The app called open but then the client got disconnected. So we shouldn't throw InvalidOperationException, we should throw a service exception with a NetworkError error code.

Reply:
Which would be retryable, and automatically retried.

Reply:
`InvalidOperationException` was thrown [here](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/service/src/Messaging/MessagesClient.cs#L414), so just throwing a retriable `IotHubServiceException` would not help and I also validated this. I am wondering calling open within "MessagesClient" instead of throwing exceptions.

Reply:
1. We never want to force customers to match exception text to know what's up and act on it in a meaningful way programmatically.
2. If the user called open, and later we are disconnected, we must not throw InvalidOperationException. Throwing that is for helping someone get their app written right, not for dealing with disconnects. Let's fix that disconnect experience to either retry connection or throw an IotHubServiceException with a NetworkError code.

@abhipsaMisra you can probably weigh in here on how to refactor this considering what you're doing for device client.